controllers/snapshotenvironmentbinding_controller.go

```diff
@@ -0,0 +1,194 @@
+	}
+	if finalizationResult.Updated {
+		if err = r.k8sClient.Update(ctx, &snapshotEnvBinding); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update snapshot env binding based on finalization result: %w", err)
+		}
+	}
+	if finalizationResult.StatusUpdated {
+		if err = r.k8sClient.Status().Update(ctx, &snapshotEnvBinding); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update the snapshot env binding status based on finalization result: %w", err)
+		}
+	}
+
+	if snapshotEnvBinding.DeletionTimestamp != nil {
+		lg.V(logs.DebugLevel).Info("SnapshotEnvironmentBinding is being deleted. skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	appNamespace := snapshotEnvBinding.Namespace
+	environmentName := snapshotEnvBinding.Spec.Environment
+	applicationName := snapshotEnvBinding.Spec.Application
+
+	// Get the Environment CR
+	environment := appstudiov1alpha1.Environment{}
+	err = r.k8sClient.Get(ctx, types.NamespacedName{Name: environmentName, Namespace: snapshotEnvBinding.Namespace}, &environment)
+	if err != nil {
+		lg.Error(err, fmt.Sprintf("unable to get the Environment %s %v", environmentName, req.NamespacedName))
+		return ctrl.Result{}, err
+	}
+
+	target, err := detectTargetFromEnvironment(ctx, environment)
+
+	remoteSecretsList := api.RemoteSecretList{}
+	if err := r.k8sClient.List(ctx, &remoteSecretsList, client.InNamespace(appNamespace), client.MatchingLabels{ApplicationLabelName: applicationName, EnvironmentLabelName: environmentName}); err != nil {
+		spiFileContentRequestLog.Error(err, "Unable to fetch remote secrets list", "namespace", appNamespace)
+		return ctrl.Result{}, unableToFetchRemoteSecretsError
+	}
+
+	for rs := range remoteSecretsList.Items {
+		remoteSecret := remoteSecretsList.Items[rs]
+		addTargetIfNotExists(&remoteSecret, target)
+		if err = r.k8sClient.Update(ctx, &remoteSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update secret: %w", err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func addTargetIfNotExists(secret *api.RemoteSecret, target api.RemoteSecretTarget) {
+	for idx := range secret.Spec.Targets {
+		existingTarget := secret.Spec.Targets[idx]
+		// Add the rest of the fields there
+		if existingTarget.Namespace == target.Namespace {
+			return
+		}
+	}
+	secret.Spec.Targets = append(secret.Spec.Targets, target)
+}
+
+func removeTarget(secret *api.RemoteSecret, target api.RemoteSecretTarget) {
+	for idx := range secret.Spec.Targets {
+		existingTarget := secret.Spec.Targets[idx]
+		// Add the rest of the fields there
+		if existingTarget.Namespace == target.Namespace {
+			secret.Spec.Targets = append(secret.Spec.Targets[:idx], secret.Spec.Targets[idx+1:]...)
+		}
+	}
+}
+
+func detectTargetFromEnvironment(ctx context.Context, environment appstudiov1alpha1.Environment) (api.RemoteSecretTarget, error) {
+	// Dummy load. Real impl should reverse pass Env -> DTC -> DT -> SpaceRequest to find out full target details
```

Comment:
We should track this as an issue.

Reply:
@metlos I am not sure if it makes sense to collect unlimited amount of duplicated objects, it will look really confusing. User's intent is not changed, user wants to have a secret in the particular namespace and controller is trying to do that. As discussed, this array is a set by its nature because it is not possible to have the same object with the same name on the same namespace and I believe we rather need to find a way to keep its contract internally than ask the external code to care of it.